Error message tutorial

Piankero opened the pull request:
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

mpranj commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,48 @@
+# Understanding Error Messages
+
+If you want to use Elektra for any of your projects you will sooner or later run into errors. These can happen
```
> If you want to use Elektra for any of your projects you will sooner or later run into errors.

I did not know we make this guarantee.

Piankero replied:
It happens quite easily for me ;) Should I change it to "you may sooner or late" ?

markus2330 replied:
Yes, the sentence is a bit weird. Simply start with the second sentence that there are many different situations where errors might occur.

mpranj replied:
Is the sentence at all relevant to understanding of error messages?

A better start for the tutorial might be.. "Errors can happen when trying to mount with insufficient permissions, [...]. The error message gives information on what the problem is." (and possibly how to resolve the problem?)

Regarding the initial sentence it might be better to open issues and give insight on how we can improve the situation. To me such a message does not seem to help the user.

Piankero commented:
@markus2330 I am not quite sure what else you expect from this tutorial.

Btw there is no shell recorder integration because it affects keys outside of `tests`. I though ran the test locally with the shell recorder and it only complained about the path being outside of `tests` (meaning everything else was ok).

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,48 @@
+Here is an example how you can trigger an error for an invalid IP entry in the `hosts` file:
+
+```sh
+sudo kdb mount --with-recommends /etc/hosts system/hosts hosts
````
Please mount to system/tests/hosts.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,48 @@
+semantic validation error a concrete subcategory of `Validation errors` (`C03XXX`) that indicate that a certain requirement is not fulfilled
+by the provided value.
+You can see all error categories along with its meanings in our [categorization guideline](../dev/error-categorization.md).
+Every category comes with its own solution approach which is intended to be as generic as possible.
```
What is a solution approach?

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,48 @@
+
+#cleanup
+sudo kdb umount system/hosts
+```
````
Now continue explaining all parts that are involved to make this error like it is and link to all deepening documents about errors (in dev and decisions).

Piankero replied:
I am still unsure what you else want. The `categorization guideline` actually contains all necessary information about the errors. I can also link to the error message format but what else do you think of?

markus2330 replied:
How to emit errors (e.g. in plugins), how to handle errors in tools, how to generate error message (-d, -v), ...

markus2330 commented:
> Btw there is no shell recorder integration because it affects keys outside of tests

Then simply put it into tests?

Piankero commented:
I added all desired changes.
What else do you think can be helpful concerning error messages? I probably am routine-blinded (betriebsblind) since I am working on this for such a long time now.

Piankero commented:
Btw I added a link which targets a specific section in another README and the linkchecker fails. I though tried it here on Github and it correctly forwards me to the specific section. Where can I disable this specific linkcheck?

markus2330 commented:
> Btw I added a link which targets a specific section in another README and the linkchecker fails. I though tried it here on Github and it correctly forwards me to the specific section. Where can I disable this specific linkcheck?

It fails because these links were not supported in some of the Markdown implementations (doxygen and/or the website). IIRC @sanssecours also stumbled over this once and investigated it. If all implementations somehow support it, it would be great to have this feature.

Piankero commented:
So what is the solution in this case? Should I remove it and just link to the whole document?

markus2330 commented:
You can try to suppress it and see if docu/webpage works.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+
+## Error message format
+
+Error messages which are emitted from Elektra's C/C++ code have the same message format for all errors
```
Why should this be specific for C/C++?

Piankero replied:
How do you say "everything non-tool related"? What I have seen is that only tools may yield other error messages

markus2330 replied:
Tools yielding other error messages are simply a bug that have not yet been fixed.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+Error messages which are emitted from Elektra's C/C++ code have the same message format for all errors
+(See [design decision](../decisions/error_message_format.md)).
+
+The design looks like this with an example reason given:
```
If it is already "like", you do not need to give "example" anymore.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+`ERROR_CODE_DESCRIPTION`: Validation of key "<key>" with string "<value>" failed.
+```
+
+If you need more information you can pass the debug argument `-d` to the command which also shows at which line in the source code
````
Wouldn't it make sense to first describe -v?

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+Configfile: /etc/hosts.13163:1575216150.11522.tmp
+```
+
+Error messages from tools (`mount`, `set`, `get`, etc.) may yield different messages such as when providing false command line arguments
````
This does not really help, linking to doc/help/kdb.md might help.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+eases the error handling. In general the [plugins README](plugins.md) is a good way to start you journey through writing your own plugin.
+All macros can be used by including `kdberrors.h` into the source file.
+For more information about the macros please read through [this](plugins.md#elektra_set_concrete_type_error) section.
+Internally when Elektra emits an error it sets a metadata to a Key which then is later on parsed into the error message.
```
This could be in more detail.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -46,3 +46,59 @@ Every category comes with its own solution approach which is intended to be as g
+```java
+		try
+		{
+			// your Elektra code
````
Which code could cause an exception?

markus2330 commented on tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ add_msr_test (
 	ENVIRONMENT KDB_EXEC_PATH=${CMAKE_SOURCE_DIR}/scripts/kdb)
 add_msr_test (tutorial_storage_plugins "${CMAKE_SOURCE_DIR}/doc/tutorials/storage-plugins.md" REQUIRED_PLUGINS type yamlcpp)
 add_msr_test (cmerge "${CMAKE_SOURCE_DIR}/doc/tutorials/cmerge.md" REQUIRED_PLUGINS hosts line PROPERTY LABELS memleak)
+add_msr_test (understanding_error_messages "${CMAKE_SOURCE_DIR}/doc/tutorials/cmerge.md" REQUIRED_PLUGINS hosts)
```
this points to the wrong file

Piankero commented:
@markus2330 I added all your desired changes, please take a look at it again

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,107 @@
+```
+
+Note that you can always pass the **command line arguments -v (verbose) and -d (debug)** to the command line to get
+extra information. Especially if you want to locate the configuration file which caused the error.
````
So what would happen in the example above?

Piankero commented:
I changed everything accordingly.
I changed the semantics of the script slightly:

1. Every line with `#` will also be written exactly like this in the tutorial
It is honestly better than echo and does not create an extra line
2. The `>>` indicate that I do an action in an opened window such as vi or kdb editor

Please review again @markus2330

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -17,8 +17,19 @@ sudo kdb set system/tests/hosts/ipv4/my 88.198.134.1777
 sudo kdb umount system/tests/hosts
 ```
 
+In this example we mount the hosts file under `/etc/hosts` into Elektra. The `hosts` plugin will also validate
````
It is not the hosts plugin itself, please look how it works (glob plugin marks with metadata, and ipaddr actually validates). This is very different to the specification model. For specifications, see doc/tutorials/validation.md

This tutorial could also cover how to specify check/validation/message.

Piankero replied:
I rewrote the sentence

Piankero commented:
@markus2330 I have added additional information on various parts, can you please look what is still missing? I really tried to incorporate everything I had in mind

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,161 @@
+sudo kdb umount system/tests/hosts
+```
+
+In this example we mount the hosts file under `/etc/hosts` into Elektra. Elektra also validates
````
Simply make assumptions (in the beginning of the tutorial) what you expect readers to know. It does not make sense to explain mounting here. The important part (why set fails), however, is not yet explained here.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+our input which is shown in the next line where we tried to set an invalid IP address. With this we can add or edit entries
+by using the tools of Elektra and do not need to edit the file directly.
+
+Note that you can always pass the **command line arguments -v (verbose) and -d (debug)** to the command line to get
```
Imho it is not so important to make this bold. It is also a bit out of the blue that this is explained. You did not yet explain why `kdb set` above fails. Alternatively: explain why we would need `-v` or `-d` here.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+
+Note that you can always pass the **command line arguments -v (verbose) and -d (debug)** to the command line to get
+extra information. Especially if you want to locate the configuration file which caused the error. Providing both parameter would
+yield an additional message which looks as following on my system:
```
Why is this dependent on your system? In which way?

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,161 @@
+extra information. Especially if you want to locate the configuration file which caused the error. Providing both parameter would
+yield an additional message which looks as following on my system:
+
+```
````
Why is this now not in shell recorder? Make one example for this whole tutorial.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,161 @@
+```
+
+You can see that every message comes with an error code which is `C03200` in the upper example which is a semantic validation error.
+Since v0.9 Elektra has hierarchically structured error codes that are leaned on to [SQL States](https://en.wikipedia.org/wiki/SQLSTATE)
````
In a tutorial it is not so important how it was. Important is how it is now. (I know anyway what your changes were). Also imho the relation to "SQL Sates" is not really important. It is explained in the design decision anyway.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+You can see that every message comes with an error code which is `C03200` in the upper example which is a semantic validation error.
+Since v0.9 Elektra has hierarchically structured error codes that are leaned on to [SQL States](https://en.wikipedia.org/wiki/SQLSTATE)
+because it is easily extensible without introducing major breaking releases. Each character has a special meaning.
+Lets take the semantic validation error `C03200` as example:
```
```suggestion
Let us take the semantic validation error `C03200` as example:
```

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+because it is easily extensible without introducing major breaking releases. Each character has a special meaning.
+Lets take the semantic validation error `C03200` as example:
+
+- `C` indicates `C`ode and should distinguish error codes from normal numbers
```
```suggestion
- `C` indicates `C`ode and distinguishes error codes from other numbers in bug reports and alike
```

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+Lets take the semantic validation error `C03200` as example:
+
+- `C` indicates `C`ode and should distinguish error codes from normal numbers
+- `03` indicates the major parent category (will be explained shortly). Two numbers were taken in case we want more than 10 major error categories in the future
```
Try to avoid forward references. Try to make one red line through everything which explains everything step by step.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+- `2` indicates a subcategory. `1` for example would be syntactic validation errors which differs in the meaning
+- `0` and `0` could be used to introduce even further subcategorizations which are under semantic validation errors
+
+Before Elektra v0.9 we had around 210 categories which were reduced to a total of 12. All macros as well as error codes can be seen in
```
Again: here it does not matter how it was. Such problems can be written in the design decisions. Maybe go over them again and update them now, so that all your documentation is consistent and nice to read.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
````diff
@@ -0,0 +1,161 @@
+that we should retry with a different value which fulfills the semantics of an IP. If we set a correct IP address, Elektra will save it to
+the `hosts` file:
+
+```
````
Again this should be in shell recorder.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+solution. Some categories of Elektra are `parent categories` of more specific error categories. E.g., is the
+semantic validation error a concrete subcategory of `Validation errors` (`C03XXX`) that indicate that a certain requirement is not fulfilled
+by the provided value. As a result some error categories are not available for putting in errors. We want to give a rough introduction
+to the three major error categories but for more detail we encourage you to look into our [categorization guideline](../dev/error-categorization.md).
```
Stress more that this should be read and by whom.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+## Error message format
+
+Error messages which are emitted from Elektra have the same message format for all errors
+(See [design decision](../decisions/error_message_format.md)).
```
Isn't the link too early. Give an overview and then say who should read this.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+For more information about the macros please read through `elektra_set_concrete_type_error` section in the [plugins.md](plugins.md) tutorial.
+Internally when Elektra emits an error it sets a [metadata](/doc/dev/metadata.md) to a Key which then is later on parsed into the error message.
+The key name of the error key is the mountpoint of the configuration. If an error occurs, 7 metadata are set which can be seen in our
+[error-handling tutorial](../dev/error-handling.md). Later on those metadata are passed into a message template which renders the message
```
In the folder "dev" there are no tutorials but technical documents. Please distinguish these documents and clearly separate this. 
https://www.divio.com/blog/documentation/

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+developers to read through our [error-message guidelines](/doc/dev/error-message.md). For out of memory errors we even enforced
+a consistent error message by altering the macro which is used for emitting errors. It does not contain a text field anymore because
+these messages should be identical in all cases. Each error message category comes with associated information which should be present.
+Validation errors for example should require the key and value in every message but is not "templated" yet. If you cannot find good possibility
```
I think this text does not help the reader of the tutorial. Ideally you would finish #3263. Alternatively, why not simply providing two templates. Then at least these parts you already did in #3263 and what new people will add can use the common layout for validation errors. Furthermore, I also think that syntax/semantic validation distinction is pretty much still unanswered in doc/dev/error-categorization.md

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+Validation errors for example should require the key and value in every message but is not "templated" yet. If you cannot find good possibility
+to add this information meaningful into the message you will most likely want to put an error into the wrong category or there is a need
+for a new category (e.g., if multiple keys are affected one might introduce a "structural validation" which accepts these). In the latter
+case please forge a [design decision](/doc/decisions/README.md) and open a pull request for discussion.
```
The tutorial should guide and give answers, not to open up new questions.

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+
+### Catching errors in language bindings
+
+Elektra provides many useful language bindings which allows you to code for Elektra in your favorite programming language.
```
The text comes out of the blue. Continue explaining the hosts errors. How would I implement what I see above in Java?

markus2330 commented on doc/tutorials/understanding-error-messages.md:
```diff
@@ -0,0 +1,161 @@
+		try (final KDB kdb = KDB.open (parentKey))
+		{
+			final KeySet ks = KeySet.create (10, KeySet.KS_END);
+            kdb.get (ks, parentKey);
```
formatting wrong

markus2330 commented:
@d3nwp maybe you can also join this discussion when you have time. In its core most of the discussion is about the distinction tutorials/design decisions/technical documents.

Would be great if we can improve the readability and targeting of our docu.

mpranj commented:
@Piankero Thank you. Unfortunately I can not merge this because it does not adhere to our code of conduct.
